adding support for calling NNLS code using Gram matrices

daniel-perry opened the pull request:
I'm interested in using your NNLS code on problems where I only have a Gram matrix, I've modified the interface and code to handle this case, and tested against solutions found with the current code.

Right now the new interface is limited to nnls algorithm, but I could extend the others if you prefer.
I tried to modify the interface as little as possible.

Let me know if you want any changes to this before a merge, or other suggestions for improvement.

(Thanks for your library, I've already found it quite useful on some problems.)

daniel-perry committed:
adding support for calling NNLS code using Gram matrices instead of data matrices.

ahwillia commented:
Thanks Daniel -- this is a great point and it should be easy to accommodate this possibility. 

I believe the `nnls_gram()` algorithm you added is actually already implemented here(?): https://github.com/ahwillia/NonNegLeastSquares.jl/blob/master/src/fnnls.jl

If this is the case (let me know if I'm mistaken -- I only took a quick look), I think it would be preferable to modify the interface along the lines you suggest to call `fnnls` directly (this way we avoid adding any redundant code in the `nnls_gram` file).

I'm glad you've found this useful, and appreciate the feedback.

daniel-perry committed:
adding Gram option to interface.

daniel-perry committed:
adding Gram option to interface.

daniel-perry committed:
updating readme to reflect Gram interface.

daniel-perry commented:
Yes, you're right fnnls is nnls with Gram matrices.
I assumed since it was a whole other paper it was doing some kind of fast approximation.
After taking a look at the paper, it appears that "fast" refers to the algorithm being faster for high dimensional data when using a Gram matrix instead of the data matrix directly.

I've modified the pull request to remove the nnls_gram.jl and tests, and now I've only changed the interface and added some tests to confirm the result is the same for data matrix vs Gram matrix call.

I also made is so that nonneg_lsq(A,b;alg=:nnls,Gram=true) does the same thing as nonneg_lsq(A,b;alg=:fnnls,Gram=true).

I've also updated the readme to reflect the new interface.

Again please advise if you'd like anything changed.

ahwillia commented:
This looks good to me. Thanks!
